File: src/lib/khqr.util.ts
/**
 * Bakong KHQR Generator Utility (EMVCo Standard)
 */

interface KHQRConfig {
    amount: number;
    currency: 'USD' | 'KHR';
    merchantName: string;
    bakongId: string;
    merchantCity?: string;
    billNumber?: string;
}

/**
 * Formats a single EMVCo Tag-Length-Value entry
 */
const formatTLV = (tag: string, value: string): string => {
    const length = value.length.toString().padStart(2, '0');
    return `${tag}${length}${value}`;
};

/**
 * CRC-16 CCITT (XModem) Calculation
 */
const calculateCRC16 = (data: string): string => {
    let crc = 0xFFFF;
    for (let i = 0; i < data.length; i++) {
        crc ^= data.charCodeAt(i) << 8;
        for (let j = 0; j < 8; j++) {
            if (crc & 0x8000) {
                crc = (crc << 1) ^ 0x1021;
            } else {
                crc = crc << 1;
            }
        }
    }
    return (crc & 0xFFFF).toString(16).toUpperCase().padStart(4, '0');
};

/**
 * Generates a real KHQR string based on EMVCo specifications
 */
export const generateKHQR = ({
    amount,
    currency,
    merchantName,
    bakongId,
    merchantCity = 'Phnom Penh',
    billNumber = '001'
}: KHQRConfig): string => {
    // 00. Payload Format Indicator
    let khqr = formatTLV('00', '01');

    // 01. Point of Initiation Method (12 = Dynamic QR)
    khqr += formatTLV('01', '12');

    // 29. Bakong Merchant Account Information
    // This is the essential part for Bakong/KHQR interoperability
    const merchantAccountInfo =
        formatTLV('00', bakongId) +
        formatTLV('01', merchantName);

    khqr += formatTLV('29', merchantAccountInfo);

    // 52. Merchant Category Code
    khqr += formatTLV('52', '0000');

    // 53. Transaction Currency (840 = USD, 116 = KHR)
    khqr += formatTLV('53', currency === 'USD' ? '840' : '116');

    // 54. Transaction Amount
    khqr += formatTLV('54', amount.toString());

    // 58. Country Code
    khqr += formatTLV('58', 'KH');

    // 59. Merchant Name
    khqr += formatTLV('59', merchantName);

    // 60. Merchant City
    khqr += formatTLV('60', merchantCity);

    // 62. Additional Data Field (Bill Number)
    const additionalData = formatTLV('01', billNumber);
    khqr += formatTLV('62', additionalData);

    // 63. CRC-16 (Checksum)
    khqr += '6304'; // Tag 63, Length 04
    khqr += calculateCRC16(khqr);

    return khqr;
};

File: android/app/src/main/java/com/myshop/cafe/utils/KhqrUtil.kt
package com.myshop.cafe.utils

import android.graphics.Bitmap
import android.graphics.Color
import com.google.zxing.BarcodeFormat
import com.google.zxing.MultiFormatWriter
import com.google.zxing.common.BitMatrix
import java.security.MessageDigest

object KhqrUtil {

    data class KhqrConfig(
        val amount: Double,
        val currency: String = "USD", // USD or KHR
        val merchantName: String,
        val accountNumber: String,
        val merchantCity: String = "PHNOM PENH",
        val billNumber: String? = null
    )

    private fun formatTlv(tag: String, value: String): String {
        val length = value.length.toString().padStart(2, '0')
        return "$tag$length$value"
    }

    /**
     * CRC16-CCITT implementation for KHQR
     */
    private fun calculateCrc16(data: String): String {
        var crc = 0xFFFF
        for (char in data) {
            val byte = char.code
            crc = crc xor (byte shl 8)
            for (i in 0 until 8) {
                if ((crc and 0x8000) != 0) {
                    crc = (crc shl 1) xor 0x1021
                } else {
                    crc = crc shl 1
                }
            }
        }
        return (crc and 0xFFFF).toString(16).uppercase().padStart(4, '0')
    }

    /**
     * Generates a KHQR string following the BSTHEN pattern (Manual Tag Builder)
     */
    fun generateKhqr(config: KhqrConfig): String {
        var khqr = ""

        // Tag 00: Payload Format Indicator
        khqr += formatTlv("00", "01")

        // Tag 01: Point of Initiation Method (11 for Static, 12 for Dynamic)
        val isDynamic = config.amount > 0
        khqr += formatTlv("01", if (isDynamic) "12" else "11")

        // Tag 29: Merchant Account Information (Individual)
        val tag29Value = formatTlv("00", config.accountNumber)
        khqr += formatTlv("29", tag29Value)

        // Tag 52: Merchant Category Code
        khqr += formatTlv("52", "5999")

        // Tag 53: Transaction Currency (840 for USD, 116 for KHR)
        khqr += formatTlv("53", if (config.currency == "USD") "840" else "116")

        // Tag 54: Transaction Amount
        if (isDynamic) {
            khqr += formatTlv("54", String.format("%.2f", config.amount))
        }

        // Tag 58: Country Code
        khqr += formatTlv("58", "KH")

        // Tag 59: Merchant Name
        khqr += formatTlv("59", config.merchantName)

        // Tag 60: Merchant City
        khqr += formatTlv("60", config.merchantCity)

        // Tag 62: Additional Data Field Template
        var tag62Value = ""
        if (config.billNumber != null) tag62Value += formatTlv("01", config.billNumber)
        tag62Value += formatTlv("02", "078211599") // Mobile Number
        tag62Value += formatTlv("03", config.merchantName) // Store Label

        if (tag62Value.isNotEmpty()) {
            khqr += formatTlv("62", tag62Value)
        }

        // Final Tag 63: CRC
        khqr += "6304"
        khqr += calculateCrc16(khqr)

        return khqr
    }

    fun generateMd5(input: String): String {
        val bytes = MessageDigest.getInstance("MD5").digest(input.toByteArray())
        return bytes.joinToString("") { "%02x".format(it) }
    }

    fun generateQrBitmap(content: String, size: Int = 512): Bitmap? {
        return try {
            val bitMatrix: BitMatrix = MultiFormatWriter().encode(
                content,
                BarcodeFormat.QR_CODE,
                size,
                size
            )
            val width = bitMatrix.width
            val height = bitMatrix.height
            val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
            for (x in 0 until width) {
                for (y in 0 until height) {
                    bitmap.setPixel(x, y, if (bitMatrix[x, y]) Color.BLACK else Color.WHITE)
                }
            }
            bitmap
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}
